src/Std/Sat/AIG/If.lean

```diff
@@ -0,0 +1,278 @@
+  let res := aig.mkAndCached ⟨notDiscr, input.rhs⟩
+  let aig := res.aig
+  let rhsRef := res.ref
+  let lhsRef := lhsRef.cast <| by
+    apply AIG.LawfulOperator.le_size_of_le_aig_size (f := mkAndCached)
+    apply AIG.LawfulOperator.le_size (f := mkNotCached)
+  aig.mkOrCached ⟨lhsRef, rhsRef⟩
+
+instance : LawfulOperator α TernaryInput mkIfCached where
+  le_size := by
+    intros
+    unfold mkIfCached
+    dsimp
+    apply LawfulOperator.le_size_of_le_aig_size (f := mkOrCached)
+    apply LawfulOperator.le_size_of_le_aig_size (f := mkAndCached)
+    apply LawfulOperator.le_size_of_le_aig_size (f := mkNotCached)
+    apply LawfulOperator.le_size (f := mkAndCached)
+  decl_eq := by
+    intros
+    unfold mkIfCached
+    dsimp
+    rw [LawfulOperator.decl_eq (f := mkOrCached)]
+    rw [LawfulOperator.decl_eq (f := mkAndCached)]
+    rw [LawfulOperator.decl_eq (f := mkNotCached)]
+    rw [LawfulOperator.decl_eq (f := mkAndCached)]
+    . apply LawfulOperator.lt_size_of_lt_aig_size (f := mkAndCached)
+      omega
+    . apply LawfulOperator.lt_size_of_lt_aig_size (f := mkNotCached)
+      apply LawfulOperator.lt_size_of_lt_aig_size (f := mkAndCached)
+      omega
+    . apply LawfulOperator.lt_size_of_lt_aig_size (f := mkAndCached)
+      apply LawfulOperator.lt_size_of_lt_aig_size (f := mkNotCached)
+      apply LawfulOperator.lt_size_of_lt_aig_size (f := mkAndCached)
+      omega
+
+theorem if_as_bool (d l r : Bool) : (if d then l else r) = ((d && l) || (!d && r))  := by
+  revert d l r
+  decide
+
+@[simp]
+theorem denote_mkIfCached {aig : AIG α} {input : TernaryInput aig} :
+    ⟦aig.mkIfCached input, assign⟧
+      =
+    if ⟦aig, input.discr, assign⟧ then ⟦aig, input.lhs, assign⟧ else ⟦aig, input.rhs, assign⟧ := by
+  rw [if_as_bool]
+  unfold mkIfCached
+  dsimp
+  simp only [TernaryInput.cast, Ref_cast', id_eq, Int.reduceNeg, denote_mkOrCached,
+    denote_projected_entry, denote_mkAndCached, denote_mkNotCached]
+  congr 2
+  . rw [LawfulOperator.denote_mem_prefix]
+    rw [LawfulOperator.denote_mem_prefix]
+    . simp
+    . simp [Ref.hgate]
+  . rw [LawfulOperator.denote_mem_prefix]
+  . rw [LawfulOperator.denote_mem_prefix]
+    rw [LawfulOperator.denote_mem_prefix]
+
+namespace RefStream
+
+structure IfInput (aig : AIG α) (w : Nat) where
+  discr : Ref aig
+  lhs : RefStream aig w
+  rhs : RefStream aig w
+
+def ite (aig : AIG α) (input : IfInput aig w) : RefStreamEntry α w :=
+  let ⟨discr, lhs, rhs⟩ := input
+  go aig 0 (by omega) discr lhs rhs .empty
+where
+  go {w : Nat} (aig : AIG α) (curr : Nat) (hcurr : curr ≤ w) (discr : Ref aig)
+      (lhs rhs : RefStream aig w) (s : RefStream aig curr) : RefStreamEntry α w :=
+    if hcurr:curr < w then
+      let input := ⟨discr, lhs.get curr hcurr, rhs.get curr hcurr⟩
+      let res := mkIfCached aig input
+      let aig := res.aig
+      let ref := res.ref
+      have := by
+        apply LawfulOperator.le_size (f := mkIfCached)
```

Comment:
I have to say I don't love this pattern of shadowing a variable, reaching a goal that cannot be written down and then doing some arcane invocation that just happens to have the right type. I would prefer something like
```suggestion
      have : aig.decls.size ≤ res.aig.decls.size :=
        LawfulOperator.le_size (f := mkIfCached) ..
```

Reply:
The shadowing of the variable is a stylistic choice such that I don't violate the linearity on `AIG` ever, simply because I can only ever reference one `AIG` at a time anyways. This is used basically all throughout the bitblaster as well to ensure linearity by convention. Do you have an alternative idea for a style that makes these linearity mistakes hard to do?

Reply:
Okay, you have convinced me.
```suggestion
      have := LawfulOperator.le_size (f := mkIfCached) ..
```
should also work I think.

Reply:
Could you write something, perhaps in the module-doc, explaining this?